libdnf/comps/group/query.cpp

```diff
@@ -42,43 +42,45 @@ namespace libdnf::comps {
 
 GroupQuery::GroupQuery(const BaseWeakPtr & base) : base(base) {
     libdnf::solv::Pool & pool = get_pool(base);
-    
-    std::map<std::string, std::vector<Id>> group_map;
+
+    std::map<std::string, std::vector<Id>> available_map;
     Id solvable_id;
+    Solvable * solvable;
     std::pair<std::string, std::string> solvable_name_pair;
-    std::string groupid;
 
     // Loop over all solvables
     FOR_POOL_SOLVABLES(solvable_id) {
+        solvable = pool.id2solvable(solvable_id);
+
         // Do not include solvables from disabled repositories
         // TODO(pkratoch): Test this works
-        if (pool.id2solvable(solvable_id)->repo->disabled) {
+        if (solvable->repo->disabled) {
             continue;
         }
         // SOLVABLE_NAME is in a form "type:id"; include only solvables of type "group"
         solvable_name_pair = split_solvable_name(pool.lookup_str(solvable_id, SOLVABLE_NAME));
         if (solvable_name_pair.first != "group") {
             continue;
         }
-        // Map groupids with list of corresponding solvable_ids
-        // TODO(pkratoch): Sort solvable_ids for each groupid according to something (repo priority / repo id / ?)
-        groupid = solvable_name_pair.second;
-        if (strcmp(pool.id2solvable(solvable_id)->repo->name, "@System")) {
-            groupid.append("_available");
+
+        // Add installed groups directly, because there is only one solvable for each
+        if (strcmp(solvable->repo->name, "@System") == 0) {
+            Group group(base);
+            group.group_ids.push_back(GroupId(solvable_id));
+            add(group);
         } else {
-            groupid.append("_installed");
-        }
-        if (group_map.find(groupid) == group_map.end()) {
-            std::vector<Id> solvable_ids;
-            group_map.emplace(groupid, solvable_ids);
+            // Create map of available groups:
+            // for each groupid (SOLVABLE_NAME), list all corresponding solvable_ids
+            // TODO(pkratoch): Sort solvable_ids for each groupid according to repo id
```

Comment:
I wonder if we should resolve this TODO in the PR. It seems semi-important and might be good to do while you've got everything fresh in the head. Let's discuss it tomorrow.

Reply:
You can find the sorting in the last but one commit now.